lua/mason-null-ls/init.lua

```diff
@@ -1,140 +1,135 @@
+				('mason-null-ls.setup_handlers: Received handler for unknown null-ls source name: %s.'):format(handler),
+				vim.log.levels.WARN
+			)
 		end
+	end, _.keys(handlers))
+
+	---@param pkg_name string
+	local function get_source_name(pkg_name)
+		return Optional.of_nilable(source_mapping.package_to_null_ls[pkg_name])
 	end
-	return vim.tbl_keys(tools)
-end
 
-local setup = function(settings)
-	SETTINGS = vim.tbl_deep_extend('force', SETTINGS, settings)
-	vim.validate({
-		ensure_installed = { SETTINGS.ensure_installed, 'table', true },
-		null_ls_sources = { SETTINGS.null_ls_sources, 'table', true },
-		auto_update = { SETTINGS.auto_update, 'boolean', true },
-		automatic_installation = { SETTINGS.automatic_installation, 'boolean', true },
-	})
-end
+	local function call_handler(source_name)
+		log.fmt_trace('Checking handler for %s', source_name)
+		Optional.of_nilable(handlers[source_name]):or_(_.always(default_handler)):if_present(function(handler)
+			log.fmt_trace('Calling handler for %s', source_name)
+			local ok, err = pcall(handler, source_name)
+			if not ok then
+				vim.notify(err, vim.log.levels.ERROR)
+			end
+		end)
+	end
 
-local show = function(msg)
-	vim.schedule_wrap(print(string.format('[mason-null-ls] %s', msg)))
+	local installed_sources = _.filter_map(get_source_name, registry.get_installed_package_names())
+	_.each(call_handler, installed_sources)
+	registry:on(
+		'package:install:success',
+		vim.schedule_wrap(function(pkg)
+			get_source_name(pkg.name):if_present(call_handler)
+		end)
+	)
 end
 
-local show_error = function(msg)
-	vim.schedule_wrap(vim.api.nvim_err_writeln(string.format('[mason-null-ls] %s', msg)))
-end
+---@return string[]
+function M.get_installed_sources()
+	local Optional = require('mason-core.optional')
+	local registry = require('mason-registry')
+	local source_mapping = require('mason-null-ls.mappings.source')
 
-local function auto_get_packages()
-	local sources = {}
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').diagnostics))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').formatting))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').code_actions))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').completion))
-	sources = vim.tbl_deep_extend('force', sources, vim.tbl_keys(require('null-ls.builtins').hover))
-	local tools = lookup(removeArrayDuplicates(sources))
-	return tools
+	return _.filter_map(function(pkg_name)
+		return Optional.of_nilable(source_mapping.package_to_null_ls[pkg_name])
+	end, registry.get_installed_package_names())
 end
 
-local do_install = function(p, version, on_close)
-	if version ~= nil then
-		show(string.format('%s: updating to %s', p.name, version))
-	else
-		show(string.format('%s: installing', p.name))
+---@param filetype string | string[]
+local function is_source_in_filetype(filetype)
```

Comment:
why return a function instead of taking in 2 parameters?

Reply:
see my response within the resolved conversation below